primo/utils/clustering_utils.py

```diff
@@ -70,31 +79,29 @@ def distance_matrix(wd: WellData, weights: dict) -> np.ndarray:
         raise_exception("Feature weights do not add up to 1.", ValueError)
 
     # Construct the matrices only if the weights are non-zero
-    cn = wd.col_names  # Column names
-    coordinates = list(zip(wd[cn.latitude], wd[cn.longitude]))
-    dist_matrix = (
+    data = wd.data if list_wells is None else wd.data.loc[list_wells]
+    cn = wd.column_names  # Column names
+    coordinates = list(zip(data[cn.latitude], data[cn.longitude]))
+    dist_matrix = wt_dist * (
         haversine_vector(coordinates, coordinates, unit=Unit.MILES, comb=True)
         if wt_dist > 0
         else 0
     )
 
-    age_range_matrix = (
-        np.abs(np.subtract.outer(wd[cn.age].to_numpy(), wd[cn.age].to_numpy()))
+    # Modifying the object in-place to save memory for large datasets
+    dist_matrix += wt_age * (
+        np.abs(np.subtract.outer(data[cn.age].to_numpy(), data[cn.age].to_numpy()))
         if wt_age > 0
         else 0
     )
 
-    depth_range_matrix = (
-        np.abs(np.subtract.outer(wd[cn.depth].to_numpy(), wd[cn.depth].to_numpy()))
+    dist_matrix += wt_depth * (
+        np.abs(np.subtract.outer(data[cn.depth].to_numpy(), data[cn.depth].to_numpy()))
         if wt_depth > 0
         else 0
     )
 
-    return (
-        wt_dist * dist_matrix
-        + wt_age * age_range_matrix
-        + wt_depth * depth_range_matrix
-    )
+    return pd.DataFrame(dist_matrix, columns=data.index, index=data.index)
 
```

Comment:
To confirm that I am understanding this correctly: the index of the data frame is what allows the computation the subset of the distance matrix while maintaining the same labeling? If that is correct, this is a great way to do this!